### ISSUE TITLE
Run Theia without a backend

### ISSUE THREAD

**Issue by sdirix** [eclipse-theia/theia#12852]
![image](https://github.com/eclipse-theia/theia/assets/8998368/a7044401-2bf5-4da5-bf86-67c2f970a768)

# tl;dr;
We would like to enable a new use case for Theia applications: Support for "browser-only" Theia applications which can be executed without a backend.

Our vision is to support Theia "browser-only" applications as a first class citizen, next to the "browser" and "electron" targets. For this we adapted the Theia CLI tooling to support the new `browser-only` target and `frontendOnly` Theia extensions. Using this tooling the feature can be supported in a minimally invasive way, requiring none (or little to none) changes to existing Theia based applications.

# Check it out yourself

You can find our PR here. Every review is appreciated!

Because it's so easy to host static sites, we deployed three example applications for you to check out:
 - [Minimal browser-only application](https://minimal--browseronly-theia.netlify.app/)
 - [Including all Theia packages](https://all--browseronly-theia.netlify.app/)
 - [POC for VS Code web extension support](https://plugins--browseronly-theia.netlify.app/)
   - You can execute a `hello world` command implemented by a Theia plugin
   - By opening a Typescript file you can see the Typescript VS Code web extension at work
   - You can clone Github repositories via the `File > Import repository` action. Note that this uses the open proxy [cors.isomorphic-git.org](https://cors.isomorphic-git.org/) so it's pretty slow. This proxy is freely available so please be considerate with its usage.
   - **Note that this POC is not optimized, so it takes a while to start.**

# Motivation
Support of browser-only use cases is a value adding feature for the whole Theia ecosystem. While the existing Theia framework offers large flexibility and therefore supports a wide array of use cases, it requires the deployment and execution of a Node based backend. This often comes with a lot of operational effort and additional costs.

There are many use cases in which not the full feature set of Theia based applications are required. For example, demo and quick start applications often come with hard coded examples anyway, do not require persistent storage and can work without terminal support.

This feature is requested from time to time in the community too:
 - https://github.com/eclipse-theia/theia/issues/8368
 - https://github.com/eclipse-theia/theia/issues/6732
 - https://github.com/eclipse-theia/theia/issues/6560
 - https://community.theia-ide.org/t/does-theia-work-without-any-backend/46
 - https://community.theia-ide.org/t/would-it-be-possible-to-create-a-backend-less-theia/1545
 - https://community.theia-ide.org/t/can-i-build-theia-as-a-browser-front-end-widget-only/646
 - https://community.theia-ide.org/t/client-only-ide-for-typescript/1622

# Implementation

Theia's tooling already generates the frontend separately from the backend in the `lib/frontend` directory. So technically the frontend can already be served with any web server. However, the frontend will try to establish a web socket connection to its backend and many services will not work when their proxy counterparts are not available.

## Tooling changes

We added a new target for Theia applications: `browser-only`, e.g

```json
  "theia": {
    "target": "browser-only"
  }
```

We added a new entry for Theia extension modules: `frontendOnly`, e.g.

```json
  "theiaExtensions": [
    {
      "frontend": "lib/browser/i18n/i18n-frontend-module",
      "frontendOnly": "lib/browser-only/i18n/i18n-frontend-only-module",
      "backend": "lib/node/i18n/i18n-backend-module"
    },
  ]
```

For this target, the Theia CLI (`theia build`) will:
 - skip generating the backend
 - for `src-gen` the `frontend`/`frontendOnly` pairs are handled in the same way as `backend`/`backendElectron` are, i.e. if both are present, `frontend` will be skipped.

Using these new tools we were able to replace, mock and/or remove all services which need to connect to the Theia backend.

`theia start` is supported for browser-only targets too: It will launch the frontend with the `webpack-dev-server`.

## Implementation changes

Using the tooling described above we were able to support the new browser-only target almost without any changes to the existing code base. The current PR!!!! only contains two breaking changes: A rename of a class, and the move of some classes from the `node` directory to the `common` directory. All other changes are only contained within `frontendOnly` modules.

The following subsections describe the most important features:

### Filesystem

Since there is no native file system available in the browser, we settled on using [BrowserFS](https://github.com/jvilk/BrowserFS), as it offers a Node compatible API, is feature rich and is well supported.

We foresee that every browser-only Theia application might want to handle the filesystem in a different way, for example to exclusively bundle some hard coded files, handling persistence via webworker cache, local storage, IndexedDB or only in memory. Therefore, we added a new service, the `BrowserFSInitialization`, that can be customized to support any behavior.

In the PR, we used this service in the new "browser-only" example application to initialize the file system with some basic example JSON files.

### WebSocketConnectionProvider

The `WebSocketConnectionProvider` of Theia is used to create proxy services for Theia's RPC mechanism.

Via our proposed PR, we already support a minimal Theia without using any proxy services consisting of `@theia/core`, `@theia/filesystem`, `@theia/editor`, `@theia/getting-started`, `@theia/keymaps`, `@theia/markers`, `@theia/monaco`, `@theia/navigator`, `@theia/outline-view`, `@theia/preferences`, `@theia/property-view`, `@theia/userstorage`, `@theia/variable-resolver` and `@theia/workspace`.

To enable using all Theia packages, even if they still depend on `WebSocketConnectionProvider`, we rebind `WebSocketConnectionProvider` to a mock implementation whose connections will always timeout. This allows using all Theia packages without further changes, even though the functionality will be broken wherever it relies on the backend, e.g. the file search will never complete.

Eventually we would like to see all Theia packages supporting their full feature set (as far as possible) for the browser-only case as well. Once that is the case, we technically no longer need a mocked  `WebSocketConnectionProvider` in the browser-only frontend, however it might make sense to still keep it for 3rd-party dependencies which are not adapted yet.

#### Outlook

For performance reasons it could make sense to move some services to web workers to lessen the load on the main Javascript thread of the frontend. For this use case an alternative `ConnectionProvider` could be implemented which bridges the gap to the web worker instead of to the backend.

# Open Tasks

With our PR!!! we think it already makes sense to merge this feature into the Theia code base. It's already possible to create specific feature-complete browser-only applications for certain use cases. The situation will only become better with more features being iteratively enabled over time. In communication with the community it must be made clear that a "browser-only" Theia can't support all use cases of a regular "browser" or "electron" Theia and that for now this feature shall be considered experimental.

We collected a non-exhaustive list of tasks which can be tackled as a follow up of the current PR.

**File management**
- Finish the BrowserFS filesystem provider
   - Implement file watching
   - Implement directory deletion
   - Implement file renaming
- Missing browser-only feature: Import Files into workspace (context menu > Upload files…)
- Missing browser-only feature: Download files from workspace (context menu > Download)
- Missing browser-only feature: Search in files

**Features**
- Support Theia plugins and VS Code web extensions
  - for more information see the following section
- Support user preferences
  - Workspace preferences are already supported via BrowserFS filesystem
- Adapt remaining pieces of preloader 
- Support terminal / console
- Support localization
- Support source control
- Support debugging
- Support OpenVSX registry
- Support tasks
  - Or remove feature if not applicable
- Show list of extensions in about dialog
- Support log levels
- Support `theia test` for browser-only targets

**Documentation**
- Add documentation for browser/browser-only modules in https://theia-ide.org/docs/architecture#separation-by-platform

## Theia plugins and VS Code web extensions support

The largest missing piece is Theia plugins and VS Code web extensions support.

The good news is that we already have a POC showcasing that the support of Theia plugins and VS Code web extensions is conceptually feasible:
 - You can find the hosted POC [here](https://plugins--browseronly-theia.netlify.app/)
   - You can execute a `hello world` command implemented by a Theia plugin
   - By opening a Typescript file you can see the Typescript VS Code web extension at work
   - You can clone Github repositories via the `File > Import repository` action. Note that this uses the open proxy [cors.isomorphic-git.org](https://cors.isomorphic-git.org/) so it's pretty slow. This proxy is freely available so please be considerate with its usage.
   - **Note that this POC is not optimized, so it takes a while to start.**
 - You can find the code on a branch [here](https://github.com/eclipsesource/theia/tree/browser-only-plugins-poc)

Conceptually, Theia is well prepared to support Theia plugins and VS Code web extensions in the browser-only case, as these are already executed in the [browser in a webworker](https://github.com/eclipse-theia/theia/tree/f00450f38bfee7d0b7f2c60b46e7f24202bdbb32/packages/plugin-ext/src/hosted/browser).

However, it will require some substantial effort until plugin support will be ready for production in the browser-only case:
 - All plugins are organizationally exclusively handled on the backend side, especially metadata collecting and deploying
 - Any call to the Theia API / VS Code API which is currently routed through the backend will not work in the browser-only case

Even though there is no obstacle for eventually fully supporting Theia plugins and VS Code web extensions in the browser-only case, it will require some work.

# Summary

With the support of browser-only applications we hope to open an exciting new chapter for the Theia framework. Please have a look at our PR and the hosted example applications and share your opinion below. Thank you!

**Comment by goyalyashpal** [comment]
so, this "backend-less/browser only" theia is to theia-desktop what github.dev or vscode.dev are to github.codespaces right?

**Comment by sdirix** [comment]
Yes that's a fair comparison when looking at the "github.dev web based editor". The clarification is needed as GitHub codespaces is also served via the github.dev domain.

**Comment by goyalyashpal** [comment]
this blog about [vscode-dev] lists many more (practical) scenarios where the serverless application becomes indispensible

[vscode-dev]: https://code.visualstudio.com/blogs/2021/10/20/vscode-dev#_local-development-with-cloud-tools

my ~summary~ distillate of it:
any scenario where full fledged desktop application can't be installed but a browser is available. so like:
* restricted machines: offices, schools, labs, libraries, etc
* restricted os-es: chromebook or window lite mode etc
* devices with unsupported os: e.g. mobile devices like android / ipad tablets etc

**Comment by msujew** [comment]
@goyalyashpal For these use cases, it's probably still best to have a full Theia instance running (for example in the cloud). The backend-less Theia version has other use-cases. For example, we're currently working on https://github.com/eclipse-theia/theia/issues/2842 which in conjunction with the backend-less Theia would allow users to join any workspace in the web without needing to setup a whole Theia server instance.

**Comment by erlmachinedev** [comment]
Extremely valuable contribution 🚀 I do Theia project from beginning and must admit that security considerations around backend-less option give me a way better confidence. 

You have an option to control everything you want via dedicated service (the great way for virtualization of resources).

P.S. Taking into account extensible design of the tool I would wish (as a user) to see browser-less version first (where the backend comes as an option)

**Comment by jjkavalam** [comment]
@sdirix Hi, found this while looking for a viable front end option for a desktop app I am planning to build.

https://minimal--browseronly-theia.netlify.app/# is almost exactly what I want as a starting point.

I will be really grateful if you could point me to the repository where the sources for this app can be found.

**Comment by msujew** [comment]
@jjkavalam It's the [browser-only example app](https://github.com/eclipse-theia/theia/tree/master/examples/browser-only) found in this repo.

**Comment by jjkavalam** [comment]
@msujew Sorry for not being clear enough.

I want to extract that example out of the repo into an independent directory (basically integrate it into another project that I have).

So I am doing:

```
mv eclipse-theia/examples/browser-only myapp
cd myapp
yarn
```

This gives me:

```
% yarn
yarn install v1.22.22
info No lockfile found.
[1/4] 🔍  Resolving packages...
error Error: https://registry.yarnpkg.com/@theia%2fapi-samples: Not found
    at params.callback [as _callback] (/Users/jobin.kavalam/.nvm/versions/node/v20.17.0/lib/node_modules/yarn/lib/cli.js:66680:18)
```

Also, I understand that the `browser-only` project is relying on generated sources. 

How to generate them ? After generation can I just move the `browser-only` directory out as mentioned above and use it independently.

**Comment by msujew** [comment]
@jjkavalam Please take a look at our documentation on [how to build your own app](https://theia-ide.org/docs/composing_applications/). In addition to the steps outlined there, you need to set the `theia.target" setting to `browser-only` like its done in the example linked above.

**Comment by jjkavalam** [comment]
@msujew Thank you. I was able to make progress. Following [the guide](https://theia-ide.org/docs/composing_applications), i have a `package.json` like below:

```
{
  "private": true,
  "dependencies": {
    "@theia/callhierarchy": "latest",
    "@theia/file-search": "latest",
    "@theia/git": "latest",
    "@theia/markers": "latest",
    "@theia/messages": "latest",
    "@theia/mini-browser": "latest",
    "@theia/navigator": "latest",
    "@theia/outline-view": "latest",
    "@theia/plugin-ext-vscode": "latest",
    "@theia/preferences": "latest",
    "@theia/preview": "latest",
    "@theia/search-in-workspace": "latest",
    "@theia/terminal": "latest"
  },
  "devDependencies": {
    "@theia/cli": "latest"
  }
}
```

However, am interested in a very minimal setup similar to https://minimal--browseronly-theia.netlify.app/.

I wonder which packages i could safely remove without breaking anything else ?

**Comment by msujew** [comment]
@jjkavalam You can likely get away just with `@theia/navigator` and `@theia/monaco`.

**Comment by kachurun** [comment]
Ok guys, I slightly improved the browser-only version in [this PR](https://github.com/eclipse-theia/theia/pull/16187). Here's what I did:

- It uses the opfs-worker package, which is an inlined web worker implementing all FS methods. With the web worker, we don't slow down the main thread with file operations, and this is the only way to ensure compatibility with all browsers (Safari can write to OPFS only with a "sync handler," which works only within a web worker). Also, this package contains an inlined version of the web worker, so we don't need any complicated setup to make it work.
- Implemented folder copy functionality. To be honest, I only realized it wasn’t implemented before when I created the PR.
- Implemented internal watching. Web worker instances (in case we have multiple Theia windows) notify about any changes made from the worker to a common broadcast channel, so no matter which worker made the change, all Theia workers receive the message. It doesn’t work for changes made outside the worker (like the OPFS explorer extension), but that’s a very narrow edge case. So now we can track changes between windows.

Also, with the broadcast channel, developers can receive notifications about changes from Theia and send them to their apps or back to Theia. For example, I have Theia integrated into my app as a separate layer in an iframe, and I use the broadcast channel to notify about file changes from either Theia or my app. I’m also using the same opfs-worker package in my app, so Theia can now detect any changes made by my app. This is amazing and is the main feature I need from the browser-only version.


---

Progress:
- [x] File read / write https://github.com/eclipse-theia/theia/pull/16187
- [x] Copy files / folders https://github.com/eclipse-theia/theia/pull/16187 
- [x] Streaming and readWriteClose capabilities. This will make opfs-worker and OPFSFileSystemProvider feature-complete and address potential issues with large files. https://github.com/eclipse-theia/theia/pull/16187
- [x] File download https://github.com/eclipse-theia/theia/pull/16187
- [x] File upload https://github.com/eclipse-theia/theia/pull/16187

- [x] Search in workspace https://github.com/eclipse-theia/theia/pull/16231
- [x] Find in files https://github.com/eclipse-theia/theia/pull/16231

**Comment by kachurun** [comment]
Search in the workspace works now (https://github.com/eclipse-theia/theia/pull/16231).
File downloading works as well.

Also, I'm going to update the outdated code from this PR (https://github.com/eclipse-theia/theia/pull/14806) to make Find in Files work.

**Comment by kachurun** [comment]
@sdirix 
I'm working on file uploading now.

Could you please help me understand why FileUploadService (frontend service) doesn't use FileService directly for file creation? Instead, it fetch `/file-upload`, which is handled by NodeFileUploadService and uses native Node fs (via fs-extra), bypassing `FileService`.

It seems to me we could just remove all of that and write files directly via FileService from FileUploadService. `FileService` depends on `FileSystemProvider`, which I have replaced in the browser only, so in that case uploading would just work, and we could remove like 2 or 3 files from the code and reduce its complexity.

But maybe I missed something?

**Comment by sdirix** [comment]
Hi @kachurun, 

If rewriting the `FileUploadService` to be based on the `FileService` is straightforward, without losing any features, then I don't think there is any reason against that. However I'm not familiar enough with the Browser File/Form APIs to evaluate the effort nor potential downsides.

At minimum, any such rewrite should be as memory efficient as the current solution as well as keep progress tracking.

What's always possible is to have a specialized implementation of `FileUploadService` for the `browser-only` case. To enable this we introduced the `browser-only` namespace. So the best way forward would be to just implement this refactoring, for now limited to `browser-only`. If it turns out that it works as well, or even better, than the original specialized implementation we can easily replace it in all scenarios.

**Comment by kachurun** [comment]
Find In Files is implemented in the same branch as Search in Files. They share some components and logic. https://github.com/eclipse-theia/theia/pull/16231

I didn't use code from [this PR](https://github.com/eclipse-theia/theia/pull/14806/files#diff-9226772e6665420220949f23071904dcd050d195a6ad284e82e0f664a1df73b3) though.

**Comment by kachurun** [comment]
> What's always possible is to have a specialized implementation of FileUploadService for the browser-only case.

Yes, that's what I did now.
Now it works—I can drop files or folders into the file tree to upload them.

**Comment by kachurun** [comment]
> What's always possible is to have a specialized implementation of FileUploadService for the browser-only case.

Yes, that's what I did.

> At minimum, any such rewrite should be as memory efficient as the current solution as well as keep progress tracking.

In my first implementation, I read into a buffer and then wrote that buffer. As a result, uploading a 500MB file used +1500MB of memory 😄

Now I rewrote it and I'm converting the native ReadableStream from File -> Theia ReadableStream and just passing it to FileService.writeFile. It uses this stream effectively if the file provider supports streaming (which I implemented for browser-only as well). Also, I made it backpressure-aware, so if the FileProvider can't consume the stream as fast as we read the file, the reading will be paused. Also, FileUploadService in browser-only no longer contains legacy methods or any of the index* methods because browser-only support is limited to Safari 15.2 / Firefox 111, so maintaining compatibility with older versions doesn’t make sense. The code is now super simple - `uploadFile` itself actually contains only one line of code.

**Comment by kachurun** [comment]
@sdirix Hi! Can you please mark everything as done in the filesystem section, thank you!

**Comment by kachurun** [comment]
> Support Theia plugins and VS Code web extensions

Well, I’m back to integrating Theia in my project, and I found out that VSIX extensions don’t work in browser-only mode for some reason, but I need themes/icon themes and languages. So it looks like I need to do something about this issue now.

So, guys, should I know anything specific about it? I hope it’s an achievable task.

**Comment by kachurun** [comment]
@sdirix 
https://github.com/eclipse-theia/theia/pull/17122
Can you please take a look at it? Thank you!

**Comment by kachurun** [comment]
> Show list of extensions in about dialog

Also fixed this
